### DISCUSSION TITLE
Google OAuth sign return null user in Flutter

### DISCUSSION THREAD

**Discussion by jagnito-studio** [supabase/supabase#9793] [Questions]
Hi,
I have a login form that uses email and google login, and use go router to always redirect non-logged in user to the login page. However, when I use the following statement for google auth, it can show up the google login form, and can redirect to my app homepage fine, but the user and session are null. 

```
TextButton(
  onPressed: () async {
    var response = await Supabase.instance.client.auth.signInWithProvider(Provider.google);
  },
  child: Text('Login with Google'),
),
```

Additionally,  
1. The google user is successfully added to the auth user table. 
2. Accessing the link "https://<my app>.supabase.co/auth/v1/authorize?provider=google" directly in browser can redirect to my app homepage as well
3. Login with email and password is fine.
4. My app has changed to use the no-hash url strategy as I checked from other discussions. Still no luck.
5. I am using package: supabase_flutter: ^0.3.3

Anything else I may have missing?

Thanks

**Comment by dshukertjr** [comment]
Just to double check, you have enabled email confirmation for email and password login, and clicking the confirmation email properly opens up your app and the session exists when you do so?

**Reply by jagnito-studio** [reply]
Hi,
Originally I have email confirmation disabled. But the email confirmation url has the same problem too - it just redirects back to the login screen due to null session. 

So am I (or the go_router that I am using) missing sth to parse the #access_token when it calls back my app's homepage?

Thanks

**Reply by dshukertjr** [reply]
@jagnito-studio 
>My app has changed to use the no-hash url strategy as I checked from other discussions. Still no luck.

This sounds like where the problem could be. Supabase actually needs the hash, so you want to turn off the no-hash strategy.

**Reply by jagnito-studio** [reply]
Hi,
I disable that and the problem is still the same (although I am using flutter web so disabling hash strategy is still quite important.)

Actually I can see the callback url can comeback as http://MY_APP_URL#access_token=xxxxx. However, I call getSessionFromUrl() but it still does not set the session
(Sorry I just identified this as I am new to flutter and supabase)

```
var url = window.location.href;

if (url.contains("#access_token")) {
  Supabase.instance.client.auth.getSessionFromUrl(Uri.parse(url));
}

Supabase.instance.client.auth.currentSession; // return null
```

Is there any other way I can set the session?

**Reply by dshukertjr** [reply]
@jagnito-studio 
I just tried it with Flutter web, and there seems to be something weird going on. There are two ways that you can make it work:
- deploy the app to a remote server
- copy the link address of the magic link you receive and paste it in the address bar of the original tab that Flutter web opens, and not on a new tab or window.

**Reply by jagnito-studio** [reply]
Hi
So far I only tried google login, not the magic link. Everything is done on the same tab only.

So I deploy it to firebase hosting, and updated my site base URL in Supabase admin to https://MY_SITE.com/ (and also https://MY_SITE.com/#/ just in case). The same problem happens when I tried google login for a new user. (But weird that the user is in the auth table). Then I press the google login button again, and the app throws an exception saying "Exception: Match error found during build phase Exception: no routes for location: access_token=...". The current URL in the browser is https://MY_SITE/#access_token=xxx. 

It sounds to me that we are missing sth to parse the access_token url??

Thanks

**Reply by dshukertjr** [reply]
@jagnito-studio Could you share your deployed app?

**Reply by jagnito-studio** [reply]
Hi, it is deployed under flutter.jagnito.com, and the git repo is https://github.com/jagnito-studio/flutter-expense

**Reply by dshukertjr** [reply]
@jagnito-studio Thanks for sharing! Let me see if I can get it to work!

**Reply by dshukertjr** [reply]
Hmm, it seems like I was able to sign in properly, but the UI hasn't updated to reflect my sign in status. 

<img width="1388" alt="Screen Shot 2022-10-28 at 21 20 04" src="https://user-images.githubusercontent.com/18113850/198585538-3200cb30-561e-4f21-adc7-fb950b3963af.png">

One thing I noticed from your code was this piece in `main.dart` file. 

```dart
var url = window.location.href;

if (url.contains("#access_token")) {
  Supabase.instance.client.auth.getSessionFromUrl(Uri.parse(url));
}
```

supabase-flutter actually takes care of getting the sessino from URL, so you won't have to run `getSessionFromUrl` manually. Sorry for the confusion.

**Reply by jagnito-studio** [reply]
Hi,
Thanks for testing it out. Yeah, that is what I observed as well, I can receive the callback with the access_token (And the login record is added to the supabase database as well). However, the session is failed to populate so it is always null. 

As I mentioned, if you press "Login with Google" again, you can see an error saying "Exception: Match error found during build phase Exception: no routes for location: access_token=...". That is why I thought something is missed to parse the URL with access_token...?

Thanks,

**Reply by dshukertjr** [reply]
@jagnito-studio 
>However, the session is failed to populate so it is always null.

Where in the code are you observing this? As the log says in my screenshot says, the user is signed in and session is not null.

>As I mentioned, if you press "Login with Google" again, you can see an error saying "Exception: Match error found during build phase Exception: no routes for location: access_token=...". That is why I thought something is missed to parse the URL with access_token...?

I have a feeling this exception is thrown during when you parse the URL. Could you remove the following code and try it?

```dart
var url = window.location.href;

if (url.contains("#access_token")) {
  Supabase.instance.client.auth.getSessionFromUrl(Uri.parse(url));
}
```

**Reply by dshukertjr** [reply]
@jagnito-studio 
Also, what would happen if you add an empty await within your [go_router method](https://github.com/jagnito-studio/flutter-expense/blob/master/lib/main.dart#L63)? Would this fix the issue?
```dart
redirect: (context, state) async {
  await Future.delayed(Duration.zero);
```

**Reply by jagnito-studio** [reply]
Hi,
> Where in the code are you observing this (null session)?

I put a breakpoint in the go router redirect that checks the currentUser and do login page/main page redirect, i.e. the following statement
```
if (state.subloc != '/login' && Supabase.instance.client.auth.currentUser == null) {
  return '/login';
} else if (state.subloc == '/login' && Supabase.instance.client.auth.currentUser != null) {
  return '/';
}
```

For you suggestion, I tried but it still fails. However, I see sth interesting

1. It once succeeded when I am in middle of local testing but using the remote hosting callback. At that time, it seems the callback url is https://MY_REMOTE_URL/#/#access_token=xxx (note the double # here). However, when I deploy it to the remote site and test it from there, it never happens again - The callback URL is just always https://MY_REMOTE_URL/#access_token=xxx
2. I keep looking at the browser and can see the URL change from sth like https://MY_REMOTE_URL -> https://MY_REMOTE_URL/#/ -> https://MY_REMOTE_URL/ -> https://MY_REMOTE_URL/#access_token.
Seems some redirect is happening...?

Thanks

**Reply by dshukertjr** [reply]
@jagnito-studio 
It seems like in your case is that the user is properly signed in, but the sign in happens after the code goes through the go router. You could use onAuthStateChange to handle the state change.

I thought this sign in happening after app launch was a counter intuitive behavior, so created a PR to fix it so that your code would work without modifying anything. If you could wait for a few days until a new version of gotrue is released with this update, it would be awesome!
https://github.com/supabase-community/supabase-flutter/pull/262

**Reply by jagnito-studio** [reply]
Thanks @dshukertjr !
